File: targets/widget/Format.swift
import Foundation

// Swift port of a SUBSET of evcc's web formatter — keep in sync with:
// https://github.com/evcc-io/evcc/blob/master/assets/js/mixins/formatter.ts
// Mirror the structure/names so it can be re-applied from source on updates.
// Numbers use the device locale (the web uses the active i18n locale).

enum PowerUnit: String {
  case w = "W"
  case kw = "kW"
  case mw = "MW"
  case auto = ""
}

private let CURRENCY_SYMBOLS: [String: String] = [
  "AUD": "$", "BGN": "лв", "BRL": "R$", "CAD": "$", "CHF": "Fr.", "CNY": "¥",
  "CZK": "Kč", "EUR": "€", "GBP": "£", "HUF": "Ft", "ILS": "₪", "JPY": "¥",
  "NZD": "$", "NOK": "kr", "PLN": "zł", "RON": "lei", "USD": "$", "DKK": "kr",
  "SEK": "kr", "ZAR": "R", "TRY": "₺", "MYR": "RM",
]

// currencies where the energy price is shown in subunits (factor 100)
private let ENERGY_PRICE_IN_SUBUNIT: [String: String] = [
  "AUD": "c", "BGN": "st", "BRL": "¢", "CAD": "¢", "EUR": "ct", "GBP": "p",
  "ILS": "ag", "NZD": "c", "NOK": "øre", "PLN": "gr", "USD": "¢", "DKK": "øre",
  "SEK": "öre", "ZAR": "c", "TRY": "krş",
]

enum Format {
  private static var locale: Locale { .current }

  private static func number(_ value: Double, _ decimals: Int, max: Int? = nil) -> String {
    let f = NumberFormatter()
    f.locale = locale
    f.numberStyle = .decimal
    f.minimumFractionDigits = decimals
    f.maximumFractionDigits = max ?? decimals
    return f.string(from: NSNumber(value: value)) ?? "\(value)"
  }

  static func energyPriceSubunit(_ currency: String) -> String? {
    if currency == "CHF" {
      return locale.language.languageCode?.identifier == "de" ? "Rp." : "ct."
    }
    return ENERGY_PRICE_IN_SUBUNIT[currency]
  }

  static func fmtNumber(_ value: Double, _ decimals: Int) -> String {
    number(value, decimals)
  }

  static func fmtW(_ watt: Double, _ format: PowerUnit = .kw, withUnit: Bool = true, digits: Int? = nil) -> String {
    var unit = format
    if unit == .auto {
      if watt >= 10_000_000 { unit = .mw }
      else if watt >= 1000 || watt == 0 { unit = .kw }
      else { unit = .w }
    }
    var value = watt
    if unit == .kw { value = watt / 1000 }
    else if unit == .mw { value = watt / 1_000_000 }
    let d = digits ?? ((unit == .kw || unit == .mw || (unit != .w && watt == 0)) ? 1 : 0)
    return number(value, d) + (withUnit ? " \(unit.rawValue)" : "")
  }

  static func fmtWh(_ watt: Double, _ format: PowerUnit = .kw, withUnit: Bool = true, digits: Int? = nil) -> String {
    fmtW(watt, format, withUnit: withUnit, digits: digits) + (withUnit ? "h" : "")
  }

  static func fmtCo2Short(_ grams: Double) -> String { "\(fmtNumber(grams, 0)) g" }
  static func fmtCo2Medium(_ grams: Double) -> String { "\(fmtNumber(grams, 0)) g/kWh" }

  static func pricePerKWhDisplayFactor(_ currency: String) -> Double {
    energyPriceSubunit(currency) != nil ? 100 : 1
  }

  static func pricePerKWhUnit(_ currency: String, short: Bool = false) -> String {
    let unit = energyPriceSubunit(currency) ?? CURRENCY_SYMBOLS[currency] ?? currency
    return "\(unit)\(short ? "" : "/kWh")"
  }

  static func fmtPricePerKWh(_ amount: Double, _ currency: String = "EUR", short: Bool = false, withUnit: Bool = true) -> String {
    let value = amount * pricePerKWhDisplayFactor(currency)
    let price = number(value, 1, max: energyPriceSubunit(currency) != nil ? 1 : 3)
    return withUnit ? "\(price) \(pricePerKWhUnit(currency, short: short))" : price
  }
}

/// Splits a "<number> <unit>" string (e.g. "8,4 kW", "23,4 ct/kWh") into parts
/// so the header can render the value large and the unit small.
func splitValueUnit(_ s: String) -> (value: String, unit: String) {
  guard let r = s.range(of: " ") else { return (s, "") }
  return (String(s[..<r.lowerBound]), String(s[r.upperBound...]))
}

File: targets/android-widget/kotlin/Format.kt
package io.evcc.android.widget

import java.util.Locale

/**
 * Kotlin port of a SUBSET of evcc's web formatter (mirrors iOS Format.swift):
 * https://github.com/evcc-io/evcc/blob/master/assets/js/mixins/formatter.ts
 * Numbers use the device locale.
 */
object Format {
    private val CURRENCY_SYMBOLS = mapOf(
        "AUD" to "$", "BGN" to "лв", "BRL" to "R$", "CAD" to "$", "CHF" to "Fr.", "CNY" to "¥",
        "CZK" to "Kč", "EUR" to "€", "GBP" to "£", "HUF" to "Ft", "ILS" to "₪", "JPY" to "¥",
        "NZD" to "$", "NOK" to "kr", "PLN" to "zł", "RON" to "lei", "USD" to "$", "DKK" to "kr",
        "SEK" to "kr", "ZAR" to "R", "TRY" to "₺", "MYR" to "RM",
    )

    // currencies where the energy price is shown in subunits (factor 100)
    private val ENERGY_PRICE_IN_SUBUNIT = mapOf(
        "AUD" to "c", "BGN" to "st", "BRL" to "¢", "CAD" to "¢", "EUR" to "ct", "GBP" to "p",
        "ILS" to "ag", "NZD" to "c", "NOK" to "øre", "PLN" to "gr", "USD" to "¢", "DKK" to "øre",
        "SEK" to "öre", "ZAR" to "c", "TRY" to "krş",
    )

    private fun number(value: Double, decimals: Int, max: Int = decimals): String {
        // format with up to `max` fraction digits, at least `decimals`, in the device locale
        val s = String.format(Locale.getDefault(), "%.${max}f", value)
        if (max <= decimals) return s
        // trim trailing zeros down to `decimals`
        val sep = java.text.DecimalFormatSymbols.getInstance().decimalSeparator
        val dot = s.indexOf(sep)
        if (dot < 0) return s
        var end = s.length
        while (end > dot + 1 + decimals && s[end - 1] == '0') end--
        if (end == dot + 1) end = dot // no fraction left
        return s.substring(0, end)
    }

    private fun energyPriceSubunit(currency: String): String? {
        if (currency == "CHF") {
            return if (Locale.getDefault().language == "de") "Rp." else "ct."
        }
        return ENERGY_PRICE_IN_SUBUNIT[currency]
    }

    fun fmtNumber(value: Double, decimals: Int): String = number(value, decimals)

    /** Power in W, auto-scaled to W/kW/MW. */
    fun fmtW(watt: Double, withUnit: Boolean = true): String {
        val unit: String
        val value: Double
        val digits: Int
        when {
            watt >= 10_000_000 -> { unit = "MW"; value = watt / 1_000_000; digits = 1 }
            watt >= 1000 || watt == 0.0 -> { unit = "kW"; value = watt / 1000; digits = 1 }
            else -> { unit = "W"; value = watt; digits = 0 }
        }
        return number(value, digits) + if (withUnit) " $unit" else ""
    }

    fun fmtWh(watt: Double): String = fmtW(watt) + "h"

    fun fmtCo2(grams: Double): String = "${number(grams, 0)} g/kWh"

    fun pricePerKWhDisplayFactor(currency: String): Double =
        if (energyPriceSubunit(currency) != null) 100.0 else 1.0

    fun pricePerKWhUnit(currency: String): String {
        val unit = energyPriceSubunit(currency) ?: CURRENCY_SYMBOLS[currency] ?: currency
        return "$unit/kWh"
    }

    fun fmtPricePerKWh(amount: Double, currency: String = "EUR", withUnit: Boolean = true): String {
        val value = amount * pricePerKWhDisplayFactor(currency)
        val price = number(value, 1, max = if (energyPriceSubunit(currency) != null) 1 else 3)
        return if (withUnit) "$price ${pricePerKWhUnit(currency)}" else price
    }
}
